File: src/core/adapters/AppAdapter.ts
/**
 * Phone Agent - AppAdapter Interface
 * Defines the contract that every platform-specific automation module must satisfy.
 */

import { JobModel } from '../../types/job';

export interface AdapterCapabilities {
  supportsVideo: boolean;
  supportsImage: boolean;
  supportsTitle: boolean;
  supportsDescription: boolean;
  supportsHashtags: boolean;
  supportsCover: boolean;
  requiresApproval: boolean;
}

export interface AdapterResult {
  success: boolean;
  message: string;
  data?: Record<string, unknown>;
  requiresUserAction?: boolean;
  finalState?: string;
}

export interface AppAdapter {
  readonly platformId: string;
  readonly packageName: string;
  readonly displayName: string;
  readonly capabilities?: AdapterCapabilities;

  isInstalled(): Promise<boolean>;
  launch(): Promise<boolean>;
  detectReadyState(): Promise<boolean>;
  selectMedia(mediaUri: string): Promise<boolean>;
  enterCaption(caption: string): Promise<boolean>;
  enterHashtags(hashtags: string[]): Promise<boolean>;
  selectCover(coverUri: string): Promise<boolean>;
  verifyPreview(): Promise<boolean>;
  requestPublishApproval(job: JobModel): Promise<boolean>;
  publish(): Promise<AdapterResult>;
  verifyPublished(): Promise<boolean>;
  recover(lastError: string): Promise<boolean>;
  stop(): Promise<void>;
}

File: android/app/src/main/java/com/phoneagent/core/adapter/AppAdapter.kt
package com.phoneagent.core.adapter

import com.phoneagent.core.model.JobModel

data class AdapterCapabilities(
    val supportsVideo: Boolean = true,
    val supportsImage: Boolean = true,
    val supportsTitle: Boolean = true,
    val supportsDescription: Boolean = true,
    val supportsHashtags: Boolean = true,
    val supportsCover: Boolean = false,
    val requiresApproval: Boolean = true
)

data class AdapterResult(
    val success: Boolean,
    val message: String,
    val data: Map<String, String> = emptyMap()
)

interface AppAdapter {
    val platformId: String
    val packageName: String
    val displayName: String
    val capabilities: AdapterCapabilities get() = AdapterCapabilities()

    suspend fun isInstalled(): Boolean
    suspend fun launch(): Boolean
    suspend fun detectReadyState(): Boolean
    suspend fun selectMedia(mediaUri: String): Boolean
    suspend fun enterCaption(caption: String): Boolean
    suspend fun enterHashtags(hashtags: List<String>): Boolean
    suspend fun selectCover(coverUri: String): Boolean
    suspend fun verifyPreview(): Boolean
    suspend fun requestPublishApproval(job: JobModel): Boolean
    suspend fun publish(): AdapterResult
    suspend fun verifyPublished(): Boolean
    suspend fun recover(lastError: String): Boolean
    suspend fun stop()
}
